### PR TITLE
`ADD|COPY` option `--keep-ownership` 

### PR DESCRIPTION
**Pull request by Dexus**
This is the follow-up PR that includes the initial PR from #3767 and is now fulfilled with additional changes, PRs (openshift/imagebuilder/pull/227) and tests.

Please follow the initial conversations in #3767.

What are the the intention for the PR:

ADD or COPY 
- without `--chown` and `--keep-ownership` result in `root:root` for the files.
- with `--chown 123:456` result in `123:456` for the files.
- with `--keep-ownership` result in `UID:GID` for the files, same as the origin source.

Closes #3767 

⚠️ Notice:

`go.mod` needs to add a replace for `github.com/openshift/imagebuilder` as long openshift/imagebuilder/pull/227 is unmerged
```
go mod edit -replace github.com/openshift/imagebuilder=github.com/Dexus-Forks/imagebuilder
```

### CODE REVIEW

File: tests/bud.bats
Comment (edsantiago) [inline tests/bud.bats:1915]: You can't do this: first, the test may run rootless. Second, even if root, you would be chowning a file that you have no right to chown (possibly in the developer's source dir, or possibly in the rpm-managed `/usr/share/tests/buildah`.

Solution: add `skip_if_rootless` to this test, and copy the build context into `$TEST_SCRATCH_DIR`.

File: tests/bud.bats
Comment (edsantiago) [inline tests/bud.bats:1917]: Whitespace and substring checks are fragile. Also, this will break if the base image ever (for whatever reason) gains those uids/gids. I encourage you to change your containerfiles to `RUN stat -c "%u:%g"` instead, and do an exact (not substring) check.

**Review by edsantiago — CHANGES_REQUESTED** [review]
Thanks for your PR! One showstopper in the tests, one strong encouragement to rework another.

File: imagebuildah/stage_executor.go
Comment (flouthoc) [inline imagebuildah/stage_executor.go:910]:
```suggestion
				return "", nil, errors.Errorf("COPY only supports the --keep-ownership, --chmod=<permissions> --chown=<uid:gid> and the --from=<image|stage> flags")
```

File: imagebuildah/stage_executor.go
Comment (flouthoc) [inline imagebuildah/stage_executor.go:913]:
```suggestion
				return "", nil, errors.Errorf("ADD only supports the --keep-ownership, --chmod=<permissions> and the --chown=<uid:gid> flags")
```

File: imagebuildah/stage_executor.go
Comment (Dexus) [reply imagebuildah/stage_executor.go:910]: Should I also add also a Comma after `--chmod=<permissions>` I was thinking about that, but thought I keep the current format.

**Comment by flouthoc** [comment]
@Dexus Commit title for `dc90aed5127822f6b210bf2ed0e2742e98a5e797` is too long could you move PR link in commit title to description please and also please sign the commits in-order to make DCO check happy.

File: imagebuildah/stage_executor.go
Comment (edsantiago) [reply imagebuildah/stage_executor.go]: >Should I also add also a Comma after --chmod=<permissions>

Yes, and also after `--chmod=<uid:gid>`. The term is `Oxford comma` and this is an instance where it can help clarity. You can also remove the superfluous `the` and not-really-helpful `=<blah>`, simplifying to:
```
...the --chmod, --chown, --from, and --keep-ownership flags
```
Also, I realize this is a WIP, but you're going to need to squash your commits (to 2, I think) at some point. There's no need nor desire for a stack of "address review feedback" commits.

Thanks again for contributing!

Comment (Dexus) [reply imagebuildah/stage_executor.go]: Thanks for your feedback, will change all you mentioned points later.

**Comment by Dexus** [comment]
So now all tests are green! 🥳 
So, what happens next?

**Comment by Dexus** [comment]
fresh rebased

**Comment by flouthoc** [comment]
> So now all tests are green! partying_face So, what happens next?

Now we have to wait for this first: https://github.com/openshift/imagebuilder/pull/227 this PR will be rebased when imagebuilder PR is merged.

File: go.mod
Comment (TomSweeneyRedHat) [inline go.mod:49]: We'll need to fix this up once the imagebuilder merges and before we merge this PR.

**Comment by TomSweeneyRedHat** [comment]
One nit request.  If you could rename the test `Dockerfile*` files to `Containerfile*`, that would be appreciated.

File: tests/bud/copy-from-keep-ownership/Dockerfile
Comment (nalind) [inline tests/bud/copy-from-keep-ownership/Dockerfile:4]: The `--keep-ownership` option is redundant when there's a `--from`, so I don't know that we need this test.

Comment (Dexus) [reply tests/bud/copy-from-keep-ownership/Dockerfile:4]: you are right

**Comment by flouthoc** [comment]
Correct me if i am wrong i think this is on hold because of https://github.com/moby/buildkit/issues/2875 since `dockerclient` does not implements this yet and imagebuilder would end up with `no-op` for `docker` implementation.

**Pull request closed without merging**
